module-bsp/board/rt1051/bsp/eink/BspEink.cpp

```diff
@@ -0,0 +1,478 @@
+            BSP_EINK_LPSPI_DMA_RX_DMA_BASE,
+            BSP_EINK_LPSPI_DMA_RX_CH,
+            BSP_EINK_LPSPI_DMA_RX_PERI_SEL,
+            BSP_EINK_LPSPI_DMA_TX_DMAMUX_BASE,
+            BSP_EINK_LPSPI_DMA_RX_DMAMUX_BASE,
+        };
+
+        AT_NONCACHEABLE_SECTION(lpspi_master_edma_handle_t lpspiEdmaHandle);
+
+        bool isInitialized = false;
+        QueueHandle_t transferCompleteQueue;
+        SemaphoreHandle_t busySemaphore; // Suspends the task until the EPD display is busy
+
+        BspEinkDriver lpspiEdmaDriverState = {
+            &lpspiResource,
+            &lpspiEdmaResource,
+            &lpspiEdmaHandle,
+            nullptr, // Will be filled in init function
+            nullptr, // Will be filled in init function
+            0,
+            0,
+            nullptr,
+            SPICSConfig::Automatic,
+            BusyWaitEventState::Registered,
+        };
+
+        void lpspiMasterEdmaCallback(LPSPI_Type *base,
+                                     lpspi_master_edma_handle_t *handle,
+                                     status_t status,
+                                     void *userData)
+        {
+            BaseType_t xHigherPriorityTaskWoken = pdFALSE;
+
+            if (xQueueSendFromISR(transferCompleteQueue, &status, &xHigherPriorityTaskWoken) != pdPASS) {
+                // If we reached this point, something very bad happened
+                return;
+            }
+
+            /* Disable SPI for current consumption reduction */
+            LPSPI_Enable(lpspiBase, false);
+
+            portEND_SWITCHING_ISR(xHigherPriorityTaskWoken);
+        }
+    } // namespace
+
+    status_t init()
+    {
+        if (isInitialized) {
+            return kStatus_Success;
+        }
+
+        /* If busy semaphore already created - delete it */
+        if (busySemaphore != nullptr) {
+            vSemaphoreDelete(busySemaphore);
+        }
+
+        /* Create new busy semaphore */
+        busySemaphore = xSemaphoreCreateBinary();
+        if (busySemaphore == nullptr) {
+            return kStatus_Fail;
+        }
+
+        /* At the beginning the semaphore is in unknown state, it has to be clearly taken or given */
+        xSemaphoreTake(busySemaphore, 0);
+
+        /* If transfer complete queue already created - delete it */
+        if (transferCompleteQueue != nullptr) {
+            vQueueDelete(transferCompleteQueue);
+        }
+
+        /* Create new queue */
+        transferCompleteQueue = xQueueCreate(1, sizeof(std::uint32_t));
+        if (transferCompleteQueue == nullptr) {
+            vSemaphoreDelete(busySemaphore);
+            busySemaphore = nullptr;
+            return kStatus_Fail;
+        }
+
+        auto lpspi           = &lpspiEdmaDriverState;
+        lpspi->eventRegister = BusyWaitEventState::NotRegistered;
+
+        gpio = DriverGPIO::Create(static_cast<GPIOInstances>(BoardDefinitions::EINK_GPIO), DriverGPIOParams{});
+        bsp::eink::eink_gpio_configure();
```

Comment:
`bsp::eink` namespace is not needed here

Reply:
Done